File: Java/OPSJLib/src/ops/DeadlineMissedFilterQoSPolicy.java
/*
 * DeadlineMissedFilterQoSPolicy.java
 *
 * Created on den 10 oktober 2007, 15:19
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

package ops;

import java.util.Observable;
import java.util.Observer;
import java.util.Timer;
import java.util.TimerTask;

/**
 *
 * @author angr
 */
public class DeadlineMissedFilterQoSPolicy extends Observable implements FilterQoSPolicy, Runnable
{
    
    long maxSeparation;

    private boolean stopped = false;
    boolean newUpdate = false;
    
    /** Creates a new instance of DeadlineMissedFilterQoSPolicy */
    public DeadlineMissedFilterQoSPolicy(long maxSeparation)
    {
        this.maxSeparation = maxSeparation;
        
        Thread t = new Thread(this);
        t.start();
        
        
        
    }
    
    public boolean applyFilter(OPSObject o)
    {
        newUpdate = true;
        return true;
    }
    public void stop()
    {
        stopped = true;
    }

    public void run()
    {
        try
        {
            while(!stopped)
            {
                Thread.sleep(maxSeparation);
                if(!newUpdate)
                {
                    setChanged();
                    notifyObservers();   
                }
                newUpdate = false;
                
            }
        } 
        catch (InterruptedException ex)
        {
            ex.printStackTrace();
            
        }
    }
    

    
    
    
    
}

File: C#/Ops/DeadlineMissedFilterQoSPolicy.cs
///////////////////////////////////////////////////////////
//  DeadlineMissedFilterQoSPolicy.cs
//  Implementation of the Class DeadlineMissedFilterQoSPolicy
//  Created on:      12-nov-2011 09:25:29
//  Author:
///////////////////////////////////////////////////////////

using System;
using System.Threading;

namespace Ops 
{
	public class DeadlineMissedFilterQoSPolicy : Observable, IFilterQoSPolicy 
    {
		internal long maxSeparation;
		internal bool newUpdate = false;
		private bool stopped = false;

        // Creates a new instance of DeadlineMissedFilterQoSPolicy
        public DeadlineMissedFilterQoSPolicy(long maxSeparation)
        {
            this.maxSeparation = maxSeparation;

            Thread t = new Thread(new ThreadStart(Run));
            t.IsBackground = true;
            t.Start(); 
        }

        public bool ApplyFilter(OPSObject o)
        {
            newUpdate = true;
            return true;
        }

        public void Stop()
        {
            stopped = true;
        }

        public void Run()
        {
            try
            {
                while (!stopped)
                {
                    Thread.Sleep((int)maxSeparation);
                    if (!newUpdate)
                    {
                        NotifyObservers(null);
                    }
                    newUpdate = false;
                }
            }
            catch (ThreadAbortException ex)
            {
                Logger.ExceptionLogger.LogException(ex);
            }
        }

	}

}